I'm back

MemoriAeni opened the pull request:
The new code does stuff

MemoriAeni committed:
Init commit of memoria_rating.php

First version of memoria_rating.php

MemoriAeni committed:
Delete memoria_rating.php

MemoriAeni committed:
init

MemoriAeni committed:
Update memoria_rating.php

changes to rounding values

MemoriAeni committed:
Update memoria_rating.php

MemoriAeni committed:
Merge branch 'dima-stefantsov:master' into master

MemoriAeni committed:
added _memoria_rating

MemoriAeni committed:
increased weight scaling

matchup prediction weight power changed from ^1.5 to ^1.6, makes matchups with higher rating difference have stronger presence in determining match prediction

MemoriAeni committed:
expanded description

MemoriAeni committed:
resulting player rating now always int

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+
+$GLOBALS['rating_calculators']['_memoria_rating'] = [
+	'description' => '
+		by <a href="/player/125980">Memoria</a>: Rating with per-matchup rating calculation, v1
```
rating rating. You might want to change wording.

lowercase after :

also this line has to fit into middle column of https://ds-rating.com/ratings , at the moment it does not, need to make it shorter.
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/8ce5b6d3-e8fb-4efd-a303-aa9f2da091d6)

MemoriAeni replied:
will be as long as dreamless rush has.. so I guess fixed
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/131588235/c859990e-8399-46d1-8917-97f312f020d7)

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+<?php
+
+$GLOBALS['rating_calculators']['_memoria_rating'] = [
```
everything in there is already "a rating", you don't have to have "rating" in the name. Up to you though.
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/c78de97b-0769-4f3e-be43-a905a8384b5f)

dima-stefantsov replied:
also think about what you will call a second rating you may want to make, a _memoria_rating2 ?

for example, it can be called something like _memoria_matchup, then when we speak, it would sound like "my memoria matchup rating is 12345".

MemoriAeni replied:
Fair point...
I'll think about it, will upload the updated code once I'm decided

MemoriAeni replied:
decided to go with "_memoria_matchups"

dima-stefantsov replied:
You have probably missed an "a", you are currently naming it "_memori_matchups"

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+		-Just like in the main rating, people start at 2000 rating,
+		 4000 rating difference = 90% win prediction, max 100 rating per match
+  		-Rating prediction is calculated for each matchup (1v1),
+		 then it\'s combined into a final prediction 
```
I suggest spending more time on this text, make it more beautiful and consistent:

- add spaces after "-", right now it looks like this, compared to level rating for example
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/1a7c0aaf-92a7-4ca2-967b-a80ba2d4cd35)

- probably get rid of linebreak before "4000", why is it a newline if it's not a new bulletpoint? we are doing things for all kinds of device sizes, including something like
![image](https://github.com/dima-stefantsov/dsr_ratings/assets/1370909/e3134103-acf0-43c5-9db8-9bb7896dc710)
, therefore you want to have semantic linebreaks, not your grandpa-linebreaks where he thought it would look good on his specific device.

- I recommend getting rid of capital letters and periods in the end, to reduce the visual clutter. But if you insist, then at least add periods consistently, right now it's 50/50.

- a comment of "just like the main rating" may become obsolete with time, you generally don't want to write something that may become wrong without you even participating in it. Who knows if I maybe decide to make default 3000 tomorrow. Up to you.

- since you are referencing "main" rating, you may want to follow default "is game accepted" logic, with it being 30 seconds (shorter games will not appear in database at all). This is totally up to you, if your vision requires 4min filter, so be it!

MemoriAeni replied:
Fair points. Will fix up

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+	];
+
+function _memoria_rating__main(&$teams) {
+	$game_duration_seconds = floor($teams[0][0]['duration']/22.4);
```
you are making unrelated checks below. This one is only checked on line 27. Therefore this variable should be first set on line 26, right before it's used. No need to do this if you are going to exit because of wrong players numbers.

MemoriAeni replied:
will fix

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+		return;
+	}
+		
+	$player_c = count($teams[0]);
```
a bad idea to name it like that.
You are not paying for extra characters, I suggest naming it $player_count, or based on your rating logic, I would even call it $team_players_count

up to you

MemoriAeni replied:
will change to player_count

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+	$factor = 4000;
+	$ratings1 = [];
+	foreach ($teams[0] as $player) {
+        $ratings1[] = $player['rating'];
```
Fix your spacing. Always use 4 spaces per level. Do not use tabs.

MemoriAeni replied:
heh, didn't notice I had spaces there instead of tabs...
used tabs everywhere but there, but if you insist, I'll change it all to spaces

dima-stefantsov replied:
yes, I insist, 4 spaces per level.

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+    }
+	$ratings2 = [];
+	foreach ($teams[1] as $player) {
+        $ratings2[] = $player['rating'];
```
this operation is a waste of not only CPU, but the person reading it brain resources.
The only place you are using it is in ($ratings1[$i]-$ratings2[$i]), I don't see much difference with ($teams[0][$i]['rating']-$teams[1][$i]['rating'])

up to you, but I highly recommend

MemoriAeni replied:
artifact from when I was testing something.. will fix

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+	$matchup_predictions_weight_total = 0;
+	for ($i = 0; $i < $player_c; $i++) {
+		$matchup_predictions[$i] = 1-(round(100/(1 + 10**(($ratings1[$i]-$ratings2[$i])/$factor))))/100;
+		$matchup_predictions_weight[$i] = (abs(((0.5 - $matchup_predictions[$i])*3)))**1.6;
```
why 3? why ^1.6?
not required, but a comment explaining wtf is going on would be nice.

MemoriAeni replied:
Pulled those numbers out of my ass. Gives me two different ways to scale the impact on the final match win prediction.

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+		$matchup_predictions_weight[$i] = (abs(((0.5 - $matchup_predictions[$i])*3)))**1.6;
+		$matchup_predictions_weight_total += $matchup_predictions_weight[$i];
+	}
+	array_multisort($matchup_predictions_weight, SORT_DESC, $matchup_predictions);
```
looks like some complicated thing that may break any moment.
does it have to be sorted? you want to compare highest chance of a 1v1matchup of one team, with highest chance of this of other team? wouldn't whatever happened to best chances be completely reversed for pairs of worst chances? or this is why you have abs(0.5-prediction), where you turn the negative outcomes to 0?

I can't foresee wth it is going to mean gamewise at all.
if you can, a comment would be cool

also can you please replace this dangerously-looking sort-code, with some simple two calls to some simple function from https://www.php.net/manual/en/array.sorting.php ? like maybe
rsort($matchup_predictions_weight);
rsort($matchup_predictions);

MemoriAeni replied:
> $matchup_predictions_weight[$i] = (abs(((0.5 - $matchup_predictions[$i])*3)))**1.6;

Win chance ranges from 0 to 1. I want matchups with high chance FROM BOTH SIDES to have most significant impact on prediction of how the matchup will probably end, so I need to know how far from 50% all the matchups are. Then I just make the impact bigger by multiplying it and raising it to the power of X.

Because of the loop following this, yes I need it sorted. array_multisort simply sorts all arrays I select by the first array's values and using options in order (only one option from my side). I'll keep that as it is.

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+	
+	$prediction = 0;
+	$normalised_weight_mult_total = 0;
+	$j = 1;
```
give this variable a name
what's this? I've been trying to understand what name it could have for like 5+ minutes, it is a counter/index of negative matchup predictions, but it is also some multiplier/denominator... I give up.

obviously not required, but highly recommended

MemoriAeni replied:
it's just a helper for this equation

> $normalised_weight_mult = (1 / ($player_count - $i)) / (($matchup_predictions_weight_total - $normalised_weight_mult_total)/$j);

so that I can divide remaining matchup_predictions_weight depending on whether it's 2v2 or 3v3 and the player it's applies to (2nd and 3rd)
I'll rewrite this loop to be more readable (+ it fixes a tiny mistake on my part)
the above equation will become:
> if ($player_count == 2) { 
$normalised_weight_mult = $matchup_predictions_weight_total - $normalised_weight_mult_total; 
}
else {
$normalised_weight_mult = ($matchup_predictions_weight_total - $normalised_weight_mult_total)/(3-$i);
$normalised_weight_mult_total += $normalised_weight_mult;
}

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+	$j = 1;
+	for ($i = 0; $i < $player_c; $i++) {
+		if ($matchup_predictions_weight_total == 0) {
+			$normalised_weight_mult = (1/$player_c);
```
no need for (), remove it

also I can't follow the logic, where you put spaces around operations, and where you don't.

MemoriAeni replied:
line 57 is changed to something else

> also I can't follow the logic, where you put spaces around operations, and where you don't.

I'll _try_ to keep it consistent

dima-stefantsov replied:
every issue mentioned should not only be fixed on the line I've commented it to, but in every piece of code where appropriate. In this case, you still have `$diff = ($multiplier*$reward[$player['winner_team']]);` for example.

MemoriAeni replied:
fixed.. I hope

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+		if ($matchup_predictions_weight_total == 0) {
+			$normalised_weight_mult = (1/$player_c);
+		}
+		elseif ($matchup_predictions[$i] == 0) {
```
== is a weak, "smart" comparison operator. If you expect exactly number zero to appear there, you write it === 0, you should always be using ===, except situations where you don't know beforehand what to expect

MemoriAeni replied:
won't fix, doesn't work with ===

dima-stefantsov replied:
it is hard for me to understand, how comparing to 0 may fail. == converts strings, etc to a number. But if what you already have in $matchup_predictions[$i] is a number, I see no reason why it should fail. We should investigate it.

MemoriAeni replied:
I don't understand why it failed either. Nonetheless this part of code is now gone.

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+
+	foreach ($teams as &$team) {
+		foreach ($team as &$player) {
+			$multiplier = _memoria_rating__get_multiplier($player['status']);
```
if you do not expect to have different multipliers, you may replace it with `$multiplier = $player['status'] === 'win' ? 1 : -1;`

MemoriAeni replied:
For now I'll change it

dima-stefantsov commented on calculators/_memoria_rating.php:
```diff
@@ -0,0 +1,89 @@
+		foreach ($team as &$player) {
+			$multiplier = _memoria_rating__get_multiplier($player['status']);
+			$diff = ($multiplier*$reward[$player['winner_team']]);
+			$player['rating'] = (int) max(1, $player['rating'] + $diff);
```
1 is already int, player-rating is already int, max returns whatever you give it. It even have to think harder when arguments are of different type. The only nonint is diff. Cast the diff itself, like
`$player['rating'] = max(10, $player['rating'] + intval($diff));`

MemoriAeni replied:
Did it like this because you advised it in last pull req, but fair enough. Will fix

dima-stefantsov commented on calculators/memoria_rating.php:
```diff
@@ -0,0 +1,74 @@
+<?php
+
+$GLOBALS['rating_calculators']['memoria_rating'] = [
```
please commit only one rating at a time. Remove this one, let us only focus on the _memoria_rating in this pull request.

MemoriAeni replied:
love github... will remove

MemoriAeni committed:
Delete calculators/memoria_rating.php

MemoriAeni committed:
Update and rename _memoria_rating.php to _memoria_matchups.php

-changed name
-changed stuff requested in pull request comments
-added another modifier to the prediction weight formula to give 50/50 matchups some weight when calculating final prediction to win
-fixed a couple bugs

MemoriAeni committed:
slightly increased the decimal accuracy of matchup prediction values

dima-stefantsov commented on calculators/_memoria_matchups.php:
```diff
@@ -0,0 +1,73 @@
+function _memori_matchups__main(&$teams) {
+    if (count($teams[0]) != count($teams[1])) {
+        return;
+    }
```
add a linebreak below }

dima-stefantsov commented on calculators/_memoria_matchups.php:
```diff
@@ -0,0 +1,73 @@
+    ];
+
+function _memori_matchups__main(&$teams) {
+    if (count($teams[0]) != count($teams[1])) {
```
use strict comparisons, !==

every non-strict comparison must have an understandable reason

dima-stefantsov commented on calculators/_memoria_matchups.php:
```diff
@@ -0,0 +1,73 @@
+                $normalised_weight_mult = $matchup_predictions_weight_total - $normalised_weight_mult_total; 
+            }
+            else {
+            $normalised_weight_mult = ($matchup_predictions_weight_total - $normalised_weight_mult_total) / (3 - $i);
```
please maintain levels depth (add 4 spaces here)

MemoriAeni committed:
fixed problems + updated the prediction weight formula

MemoriAeni commented:
Made the ELO equation symmetrical rating-wise, and changed the matchup weight calculation so that it scales properly and there's no more anomalies where your win prediction might be higher when it should in fact be lower.

MemoriAeni committed:
removed obsolete variable

dima-stefantsov commented on calculators/_memoria_matchups.php:
```diff
@@ -0,0 +1,62 @@
+    
+    $prediction = 0;
+    for ($i = 0; $i < count($teams[0]); $i++) {
+        $normalised_weight_mult = $matchup_predictions_weight[$i] / $matchup_predictions_weight_total;
```
you start with $matchup_predictions_weight_total=0, if something goes wrong, and a team will have 0 players, the code will break, instead of simply not assigning any rating.
(edit: I see you are calling it inside a cycle that will not be executed when count==0, but still scary, a default value of 0 for something we will be using as unvalidated divisor)

then, even if teams are not empty, how do you guarantee `((abs(0.5 - floor(1000 / (1 + 10 ** (abs(($teams[0][$i]['rating'] - $teams[1][$i]['rating']) / $factor)))) / 1000) + 0.015) * 6) ** (abs(($teams[0][$i]['rating'] - $teams[1][$i]['rating']) / $factor) * 1.5)`? Right now it can be followed through formulas, and maybe your if() guarantees it (I've lost it there), but this is way too complicated to be sustainable. Even if it ends up guaranteeing it now, next time you try to improve algorithm and break this brittle balance.

Make absolutely sure it is never 0. Code must not break. Even if($total === 0) {$total = 1;} would work, although it's a hack to an algorithm.

MemoriAeni replied:
If a team has 0 players? Doesn't your server discard all such replays now? Correct me if I'm wrong... not that that would break my algorithm anyway.

As for the formula guaranteeing there won't be a 0 division error, it mathematically simply isn't possible. Not only anything to the power of 0 gives 1 and $elo_part is always positive because of the absolute value, but the value of `0.5 - $matchup_predictions[$i]` increases with higher value of `$elo_part * 1.5`, so there's no way the algorithm produces something like 0.09^15 (which would probably indeed be rounded to 0), let alone three times in a row for a 3v3 game. 40k rating difference would result in 1 winning chance, `((abs(0.5 - $matchup_predictions[$i]) + 0.015) * 6) ** (10 * 1.5)` would thus be 5.8e11. Not even remotely close to 0.
I could calculate the actual limits for the equation if I remembered how to to prove my point, but I'm not going to. Unless you ever decide to not discard unrateable games that all your other algorithms would ignore, just not mine (which I don't see how or why you'd do that), this won't be an issue.

dima-stefantsov replied:
This is just not how you should write your code. To use a divisor, that is 0, and then stars must align to keep changing it to nonzero, over chained formulas, without any precautions like `if(divisor==0) then return;`

It is also not sustainable, because right now you, it seems, indeed achieved always non-zero divisor. But a year later, you, or even worse someone else will want do adjust this code, "this +0.015 coefficient is throwing balance off, I think I'll remove it", but this was the only thing that kept the code from exploding, who would ever know?

Since you do not agree, sure, I'll accept the way it is, first time your code really kills the site, it will be easier for me to argue :)

MemoriAeni committed:
removed brackets, fixed rounding issue
